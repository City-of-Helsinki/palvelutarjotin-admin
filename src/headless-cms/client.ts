/* eslint-disable @typescript-eslint/explicit-module-boundary-types */
import { ApolloClient, InMemoryCache } from '@apollo/client';
import { relayStylePagination } from '@apollo/client/utilities';

const apolloClient = new ApolloClient({
  uri: import.meta.env.VITE_APP_CMS_URI,
  cache: new InMemoryCache({
    typePolicies: {
      Page: {
        fields: {
          children: relayStylePagination(['where', ['search']]),
        },
      },
    },
  }),
  headers: {
    authorization: '',
  },
});
export default apolloClient;
