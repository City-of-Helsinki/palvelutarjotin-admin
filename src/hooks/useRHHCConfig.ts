import type { ApolloClient, NormalizedCacheObject } from '@apollo/client';
import React from 'react';
import type { Config, ModuleItemTypeEnum } from 'react-helsinki-headless-cms';
import { defaultConfig as rhhcDefaultConfig } from 'react-helsinki-headless-cms';
import { useTranslation } from 'react-i18next';

import AppConfig from '../domain/app/AppConfig';
import { MAIN_CONTENT_ID } from '../domain/app/layout/PageLayout';
import { getCmsPath } from '../domain/app/routes/utils';
import getLanguageCode from '../utils/getLanguageCode';
import useLocale from './useLocale';

const getIsHrefExternal = (href: string) =>
  !href?.startsWith('/') &&
  !AppConfig.internalHrefOrigins.some((origin) => href?.includes(origin));

const getRoutedInternalHref = (
  link?: string | null,
  type?: ModuleItemTypeEnum
): string => getCmsPath(link ?? '');

type Props = {
  apolloClient: ApolloClient<NormalizedCacheObject>;
  eventsApolloClient: ApolloClient<NormalizedCacheObject>;
  venuesApolloClient: ApolloClient<NormalizedCacheObject>;
};

export default function useRHHCConfig({
  apolloClient,
  eventsApolloClient,
  venuesApolloClient,
}: Props): Config {
  const { t } = useTranslation();
  const locale = useLocale();

  return React.useMemo(() => {
    return {
      ...rhhcDefaultConfig,
      mainContentId: MAIN_CONTENT_ID,
      organisationPrefixes: [],
      components: {
        ...rhhcDefaultConfig.components,
      },
      siteName: t('appName'),
      currentLanguageCode: getLanguageCode(locale),
      apolloClient,
      eventsApolloClient,
      venuesApolloClient,
      utils: {
        ...rhhcDefaultConfig.utils,
        getIsHrefExternal,
        getRoutedInternalHref,
      },
      internalHrefOrigins: AppConfig.internalHrefOrigins,
      copy: {
        ...rhhcDefaultConfig.copy,
        menuToggleAriaLabel: 'toggle',
        skipToContentLabel: t('common.linkSkipToContent'),
      },
      htmlSanitizer: {
        allowedUnsafeTags: ['iframe'],
        trustedOrigins: ['https://www.youtube.com', 'https://player.vimeo.com'],
      },
    };
  }, [t, locale, apolloClient, eventsApolloClient, venuesApolloClient]);
}
