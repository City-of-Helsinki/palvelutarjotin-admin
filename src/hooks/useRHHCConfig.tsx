import type { Config } from 'react-helsinki-headless-cms';
import { defaultConfig as rhhcDefaultConfig } from 'react-helsinki-headless-cms';
import { useTranslation } from 'react-i18next';
import { Link } from 'react-router';

import useLocale from './useLocale';
import AppConfig from '../domain/app/AppConfig';
import { MAIN_CONTENT_ID } from '../domain/app/layout/PageLayout';
import { getCmsPath } from '../domain/app/routes/utils';
import { stripLocaleFromUri } from '../headless-cms/utils';
import getLanguageCode from '../utils/getLanguageCode';

const getIsHrefExternal = (href: string) =>
  !href?.startsWith('/') &&
  !AppConfig.internalHrefOrigins.some((origin) => href?.includes(origin));

type Props = Pick<
  Config,
  'apolloClient' | 'eventsApolloClient' | 'venuesApolloClient'
>;

export default function useRHHCConfig({
  apolloClient,
  eventsApolloClient,
  venuesApolloClient,
}: Props): Config {
  const { t } = useTranslation();
  const locale = useLocale();
  return {
    ...rhhcDefaultConfig,
    mainContentId: MAIN_CONTENT_ID,
    organisationPrefixes: [],
    components: {
      ...rhhcDefaultConfig.components,

      A: ({ href, ...props }) => <Link to={href ?? ''} {...props} />,

      Link: ({ href, ...props }) => <Link to={href ?? ''} {...props} />,
    },
    siteName: t('appName'),
    currentLanguageCode: getLanguageCode(locale),
    apolloClient,
    eventsApolloClient,
    venuesApolloClient,
    utils: {
      ...rhhcDefaultConfig.utils,
      getIsHrefExternal,
      getRoutedInternalHref: (link?: string | null) =>
        `/${locale.toLowerCase()}${getCmsPath(stripLocaleFromUri(link ?? ''))}`,
    },
    internalHrefOrigins: AppConfig.internalHrefOrigins,
    copy: {
      ...rhhcDefaultConfig.copy,
      menuToggleAriaLabel: 'toggle',
      skipToContentLabel: t('common.linkSkipToContent'),
    },
    htmlSanitizer: {
      allowedUnsafeTags: ['iframe'],
      trustedOrigins: ['https://www.youtube.com', 'https://player.vimeo.com'],
    },
  };
}
