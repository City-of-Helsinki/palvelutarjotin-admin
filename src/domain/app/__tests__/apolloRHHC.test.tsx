import { vi } from 'vitest';
import {
  ApolloClient,
  NormalizedCacheObject,
  useApolloClient,
} from '@apollo/client';
import { MockedProvider } from '@apollo/client/testing';
import { render, renderHook } from '@testing-library/react';
import React from 'react';
import { ConfigProvider, useConfig } from 'react-helsinki-headless-cms';
import {
  LanguagesDocument,
  MenuDocument,
  Navigation,
  useLanguagesQuery,
} from 'react-helsinki-headless-cms/apollo';

import useRHHCConfig from '../../../hooks/useRHHCConfig';
import {
  customRender,
  RHHCConfigProviderWithProvidedApolloClient,
} from '../../../utils/testUtils';

const mocks = [
  {
    request: { query: LanguagesDocument, variables: {} },
    result: { data: { languages: [] } },
  },
  {
    request: {
      query: MenuDocument,
      variables: { menuIdentifiersOnly: false, id: '' },
    },
    result: { data: { menu: null } },
  },
];

describe('test issues in connection between the apollo client and the RHHC-lib', () => {
  describe('apollo client should work with RHHC-lib', () => {
    it('has apollo client initialized properly in the config', () => {
      const { result } = renderHook(() => useConfig());
      expect(result.current.apolloClient).not.toBeNull();
    });

    it('renders any component that needs the apollo client', () => {
      const { container } = customRender(
        <Navigation
          onTitleClick={function (): void {
            throw new Error('Function not implemented.');
          }}
          menuName={''}
          getPathnameForLanguage={function (): string {
            throw new Error('Function not implemented.');
          }}
        />,
        { mocks }
      );
      expect(container).toBeInTheDocument();
    });

    const ApolloClientFromUseApolloClient = () => {
      const apolloClient = useApolloClient();
      return <div>{apolloClient.version}</div>;
    };

    const ApolloClientFromUseConfig = () => {
      const { apolloClient } = useConfig();
      return <div>{apolloClient?.version}</div>;
    };

    const ApolloNavigationComponent = () => (
      <Navigation
        onTitleClick={vi.fn()}
        menuName={''}
        getPathnameForLanguage={vi.fn()}
      />
    );

    const ApolloLanguageQueryComponent = () => {
      const { data, loading } = useLanguagesQuery();
      if (loading || !data) {
        return <div data-testing-id="loading-spinner">LOADING...</div>;
      }
      return (
        <div data-testing-id="languages-query-data">
          data: {JSON.stringify(data)}
        </div>
      );
    };

    it.each([
      ApolloClientFromUseApolloClient,
      ApolloClientFromUseConfig,
      ApolloNavigationComponent,
      ApolloLanguageQueryComponent,
    ])(
      'renders component that needs the apollo client with a custom wrapper %o',
      (Component) => {
        const { container } = render(<Component />, {
          wrapper: ({ children }) => (
            <MockedProvider mocks={mocks}>
              <RHHCConfigProviderWithProvidedApolloClient>
                {children}
              </RHHCConfigProviderWithProvidedApolloClient>
            </MockedProvider>
          ),
        });
        expect(container).toMatchSnapshot();
      }
    );
  });

  describe('mockedProvider', () => {
    it('offers apollo client', () => {
      const { result } = renderHook(() => useApolloClient(), {
        wrapper: ({ children }) => (
          <MockedProvider mocks={mocks}>{children}</MockedProvider>
        ),
      });
      expect(result.current).not.toBeNull();
    });
    it('apollo client is usable with useConfig', () => {
      const {
        result: { current: apolloClient },
      } = renderHook(() => useApolloClient(), {
        wrapper: ({ children }) => (
          <MockedProvider mocks={mocks}>{children}</MockedProvider>
        ),
      });
      const { result } = renderHook(() =>
        useRHHCConfig({
          apolloClient: apolloClient as ApolloClient<NormalizedCacheObject>,
        })
      );
      expect(result.current.apolloClient).not.toBeNull();
    });
    describe('useRHHCConfig', () => {
      it('works with apollo provider', () => {
        const {
          result: { current: apolloClient },
        } = renderHook(() => useApolloClient(), {
          wrapper: ({ children }) => (
            <MockedProvider mocks={mocks}>{children}</MockedProvider>
          ),
        });
        const {
          result: { current: rhhcConfig },
        } = renderHook(() =>
          useRHHCConfig({
            apolloClient: apolloClient as ApolloClient<NormalizedCacheObject>,
          })
        );
        const wrapper = ({ children }: { children: React.ReactNode }) => (
          <ConfigProvider config={rhhcConfig}>{children}</ConfigProvider>
        );

        render(
          <Navigation
            onTitleClick={function (): void {
              throw new Error('Function not implemented.');
            }}
            menuName={''}
            getPathnameForLanguage={function (): string {
              throw new Error('Function not implemented.');
            }}
          />,
          { wrapper }
        );
      });
    });
  });
});
