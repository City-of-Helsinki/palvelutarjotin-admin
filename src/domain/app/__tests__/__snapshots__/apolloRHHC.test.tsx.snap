// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`test issues in connection between the apollo client and the RHHC-lib > apollo client should work with RHHC-lib > renders component that needs the apollo client with a custom wrapper [Function ApolloClientFromUseApolloClient] 1`] = `
<div>
  <div>
    3.8.1
  </div>
</div>
`;

exports[`test issues in connection between the apollo client and the RHHC-lib > apollo client should work with RHHC-lib > renders component that needs the apollo client with a custom wrapper [Function ApolloClientFromUseConfig] 1`] = `
<div>
  <div>
    3.8.1
  </div>
</div>
`;

exports[`test issues in connection between the apollo client and the RHHC-lib > apollo client should work with RHHC-lib > renders component that needs the apollo client with a custom wrapper [Function ApolloLanguageQueryComponent] 1`] = `
<div>
  <div
    data-testing-id="loading-spinner"
  >
    LOADING...
  </div>
</div>
`;

exports[`test issues in connection between the apollo client and the RHHC-lib > apollo client should work with RHHC-lib > renders component that needs the apollo client with a custom wrapper [Function ApolloNavigationComponent] 1`] = `<div />`;
